# Escaleras y serpientes (tarea 2 RL)
Juan Pablo Reyes Fajardo

Santiago Rodríguez Ávila

<center>
    <img src="images/tablero.png" alt="centered image" width=300/>
</center>

Se tiene el siguiente juego de escalera, hay serpientes que hacen descender y escaleras que hacen ascender. Los puntos rojos y azules son casillas en donde se va a terminar el juego, si cae en la casilla roja se pierde la partida, si cae en la casilla azul se gana la partida.

La idea es utilizar diferentes métodos de aprendizaje por refuerzo para ver cómo se comporta con el juego.

### Reglas del juego

* El jugador decide si avanzar o retroceder la cantidad de casillas que indique el dado **Antes** de lanzarlo.
* Si se llega a uno de los bordes de la escalera (1 o 100), el jugador \"rebotará\" en las casillas y retrocederá o avanzará (dependiendo el caso) el número de casillas restantes."

## Modelando el MDP

* Estados -> 100. Estos corresponden a las posiciones posibles del tablero. Cuando el agente llega al inicio de una escalera o en la cabeza de una serpiente, el estado al que va a seguir el agente va a ser desde el otro lado de la escalera/serpiente.
* Recompensas -> 1. si el jugador llega a un punto azul, -1 si el jugador llega a un punto rojo. -0.1 de lo contrario.
* Acciones -> 2. El agente solo se puede mover en dos direcciones: adelante (aumentando casillas) o atrás (retrocediendo)."

**Nota**: La recompensa de paso negativo se colocó para que el agente no tratara de evitar los puntos rojos, si no que se enfoque en llegar a algún punto terminal bueno. Por eso es que la recompensa de paso es un poco negativa.

A continuación se utilizará el módulo de *Python* creado por los estudiantes llamado *escalerasyserpientes.py* para crear el tablero, los puntos terminales (positivos como negativos) y el valor de la recompensa del paso:

In [2]:
from types import MethodType
from escalerasyserpientes import EscalerasSerpientes
import copy
import numpy as np

Inicializar el MDP:

In [3]:
es=EscalerasSerpientes([80,100],[23,37,45,67,89],-0.1)

Visualizar los estados:

In [58]:
print(es.states)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100]


Visualizar la política

In [59]:
print(es.policy)

{1: 'Ad', 2: 'Ad', 3: 'Ad', 4: 'Ad', 5: 'Ad', 6: 'Ad', 7: 'Ad', 8: 'Ad', 9: 'Ad', 10: 'Ad', 11: 'Ad', 12: 'Ad', 13: 'Ad', 14: 'Ad', 15: 'Ad', 16: 'Ad', 17: 'Ad', 18: 'Ad', 19: 'Ad', 20: 'Ad', 21: 'Ad', 22: 'Ad', 23: 'Ad', 24: 'Ad', 25: 'Ad', 26: 'Ad', 27: 'Ad', 28: 'Ad', 29: 'Ad', 30: 'Ad', 31: 'Ad', 32: 'Ad', 33: 'Ad', 34: 'Ad', 35: 'Ad', 36: 'Ad', 37: 'Ad', 38: 'Ad', 39: 'Ad', 40: 'Ad', 41: 'Ad', 42: 'Ad', 43: 'Ad', 44: 'Ad', 45: 'Ad', 46: 'Ad', 47: 'Ad', 48: 'Ad', 49: 'Ad', 50: 'Ad', 51: 'Ad', 52: 'Ad', 53: 'Ad', 54: 'Ad', 55: 'Ad', 56: 'Ad', 57: 'Ad', 58: 'Ad', 59: 'Ad', 60: 'Ad', 61: 'Ad', 62: 'Ad', 63: 'Ad', 64: 'Ad', 65: 'Ad', 66: 'Ad', 67: 'Ad', 68: 'Ad', 69: 'Ad', 70: 'Ad', 71: 'Ad', 72: 'Ad', 73: 'Ad', 74: 'Ad', 75: 'Ad', 76: 'Ad', 77: 'Ad', 78: 'Ad', 79: 'Ad', 80: 'Ad', 81: 'Ad', 82: 'Ad', 83: 'Ad', 84: 'Ad', 85: 'Ad', 86: 'Ad', 87: 'Ad', 88: 'Ad', 89: 'Ad', 90: 'Ad', 91: 'Ad', 92: 'Ad', 93: 'Ad', 94: 'Ad', 95: 'Ad', 96: 'Ad', 97: 'Ad', 98: 'Ad', 99: 'Ad', 100: 'Ad'}


### Revisión del módulo

Para revisar que el módulo funciona correctamente, se va a probar la función de pasos para estados resultantes, para esto se tendrá un valor de estado y un valor de lanzado de dado, también se hará el paso hacia adelante. El resultado fue el siguiente:

In [60]:
state_prueba=98
action_prueba=5
print(es.step(state_prueba, action_prueba, random=False))

(97, -0.1, 5, False)


Se puede ver que está bien implementado ya que se tiene el valor de la casilla de "rebote" una vez el agente llega a la casilla 100.

Ahora se hará una prueba para probar función de pasos para acciones aleatorias del agente (se va a lanzar un dado y ver cómo se comporta):

Estructura de retorno:
* Estado siguiente, recompensa estado, Accion tomada, Estado terminal (True/False)

In [4]:
state_prueba=98
action_prueba='Ad'
for i in range(6*3):
    print(es.step(state_prueba, action_prueba, random=True))


(28, -0.1, 4, False)
(97, -0.1, 5, False)
(96, -0.1, 6, False)
(99, -0.1, 3, False)
(28, -0.1, 4, False)
(96, -0.1, 6, False)
(100, -0.1, 2, False)
(28, -0.1, 4, False)
(99, -0.1, 1, False)
(96, -0.1, 6, False)
(28, -0.1, 4, False)
(99, -0.1, 3, False)
(28, -0.1, 4, False)
(99, -0.1, 1, False)
(99, -0.1, 1, False)
(99, -0.1, 1, False)
(96, -0.1, 6, False)
(100, -0.1, 2, False)


Nuevamente, se puede comprobar que el módulo está correctamente implementado porque se tienen los valores esperados dependiendo del valor del dado.

## Actualización de valores y programación dinámica

A continuación se implementará un algoritmo de estimación de valores que se hará con programación dinámica, luego se hará una prueba con un factor de descuento  $\gamma$ y un horizonte H determinado.

In [5]:
def update_values(es, gamma):
    # Arreglos para los valores
    value = dict.fromkeys(es.states , 0.0)
    new_values = dict.fromkeys(es.states , 0.0)
    
    for state in es.states: 
        if state in es.snakes:
            new_values[state]=es.state_values[es.snakes[state]]
        elif state in es.stairs:
            new_values[state]=es.state_values[es.stairs[state]]
        else:
            # Tomar acción de politica dada
            action = es.policy[state]
            prob=es.action_probabilities[0]
            
            real_actions=es.real_actions[action]
            for real_action in real_actions:
                # Dar un paso
                new_state, reward, _, done = es.step(state, real_action, random=False)
                # Actualizar valores
                if(done):
                    # actualización para un estado terminal
                    new_values[state] = reward
                else:
                    # actualización para un estado no terminal
                    new_values[state] += prob*(reward + gamma*es.state_values[new_state]) #borrar

    # Copiar valores
    value = copy.deepcopy(new_values) 
    es.state_values = copy.deepcopy(new_values) 

Gamma en 0.9, horizonte en 2 y política siempre adelante

In [6]:
gamma = 0.9
H = 2

Inicializa valores y corre programación dinámica

In [7]:
es.state_values=es.init_values()
es.update_values = MethodType(update_values, es)
es.solve_dynamic_programming(gamma=gamma, horizon=H)

Función de valor:

In [8]:
for state in es.states:
    print(f'V({state}) = {es.state_values[state]:.3f}')

V(1) = -0.190
V(2) = -0.190
V(3) = -0.190
V(4) = -0.190
V(5) = -0.190
V(6) = -0.190
V(7) = -0.190
V(8) = -0.100
V(9) = -0.190
V(10) = -0.190
V(11) = -0.190
V(12) = -0.190
V(13) = -0.190
V(14) = -0.190
V(15) = -0.190
V(16) = -0.190
V(17) = -0.325
V(18) = -0.325
V(19) = -0.325
V(20) = -0.325
V(21) = -0.100
V(22) = -0.325
V(23) = -1.000
V(24) = -0.190
V(25) = -0.190
V(26) = -0.190
V(27) = -0.190
V(28) = -0.190
V(29) = -0.190
V(30) = -0.190
V(31) = -0.325
V(32) = -0.325
V(33) = -0.325
V(34) = -0.325
V(35) = -0.325
V(36) = -0.325
V(37) = -1.000
V(38) = -0.190
V(39) = -0.325
V(40) = -0.325
V(41) = -0.325
V(42) = -0.325
V(43) = -0.100
V(44) = -0.100
V(45) = -1.000
V(46) = -0.100
V(47) = -0.190
V(48) = -0.100
V(49) = -0.190
V(50) = -0.100
V(51) = -0.190
V(52) = -0.100
V(53) = -0.190
V(54) = -0.100
V(55) = -0.100
V(56) = -0.190
V(57) = -0.190
V(58) = -0.190
V(59) = -0.100
V(60) = -0.190
V(61) = -0.325
V(62) = -0.100
V(63) = -0.325
V(64) = -0.100
V(65) = -0.325
V(66) = -0.100
V(67) = -1.000
V(68

Ahora se volverá a hacer la prueba pero con otra política, esta vez, aleatoria:

In [9]:
for i in es.policy: 
    rand=np.random.uniform()
    if rand<0.5:
        es.policy[i]='Ad'
    else:
        es.policy[i]='At'

In [10]:
print(es.policy)

{1: 'Ad', 2: 'At', 3: 'At', 4: 'At', 5: 'Ad', 6: 'At', 7: 'At', 8: 'At', 9: 'At', 10: 'Ad', 11: 'Ad', 12: 'Ad', 13: 'At', 14: 'At', 15: 'Ad', 16: 'At', 17: 'At', 18: 'At', 19: 'At', 20: 'At', 21: 'At', 22: 'At', 23: 'At', 24: 'At', 25: 'At', 26: 'At', 27: 'Ad', 28: 'Ad', 29: 'Ad', 30: 'At', 31: 'At', 32: 'Ad', 33: 'At', 34: 'Ad', 35: 'Ad', 36: 'Ad', 37: 'Ad', 38: 'At', 39: 'At', 40: 'At', 41: 'Ad', 42: 'Ad', 43: 'At', 44: 'At', 45: 'At', 46: 'Ad', 47: 'Ad', 48: 'Ad', 49: 'Ad', 50: 'Ad', 51: 'Ad', 52: 'At', 53: 'Ad', 54: 'Ad', 55: 'Ad', 56: 'Ad', 57: 'At', 58: 'At', 59: 'Ad', 60: 'At', 61: 'Ad', 62: 'Ad', 63: 'Ad', 64: 'Ad', 65: 'Ad', 66: 'Ad', 67: 'Ad', 68: 'At', 69: 'Ad', 70: 'Ad', 71: 'Ad', 72: 'At', 73: 'At', 74: 'Ad', 75: 'Ad', 76: 'At', 77: 'At', 78: 'Ad', 79: 'Ad', 80: 'At', 81: 'At', 82: 'Ad', 83: 'Ad', 84: 'Ad', 85: 'Ad', 86: 'Ad', 87: 'Ad', 88: 'Ad', 89: 'At', 90: 'Ad', 91: 'At', 92: 'At', 93: 'At', 94: 'Ad', 95: 'Ad', 96: 'At', 97: 'At', 98: 'Ad', 99: 'At', 100: 'At'}


In [11]:
gamma = 0.9
H = 2

In [12]:
es.state_values=es.init_values()
es.update_values = MethodType(update_values, es)
es.solve_dynamic_programming(gamma=gamma, horizon=H)

Función de valor:

In [13]:
for state in es.states:
    print(f'V({state}) = {es.state_values[state]:.3f}')

V(1) = -0.190
V(2) = -0.190
V(3) = -0.190
V(4) = -0.190
V(5) = -0.190
V(6) = -0.190
V(7) = -0.190
V(8) = -0.100
V(9) = -0.190
V(10) = -0.190
V(11) = -0.190
V(12) = -0.190
V(13) = -0.190
V(14) = -0.190
V(15) = -0.190
V(16) = -0.190
V(17) = -0.190
V(18) = -0.190
V(19) = -0.190
V(20) = -0.190
V(21) = -0.100
V(22) = -0.190
V(23) = -1.000
V(24) = -0.325
V(25) = -0.325
V(26) = -0.325
V(27) = -0.190
V(28) = -0.190
V(29) = -0.190
V(30) = -0.190
V(31) = -0.190
V(32) = -0.325
V(33) = -0.190
V(34) = -0.325
V(35) = -0.325
V(36) = -0.325
V(37) = -1.000
V(38) = -0.325
V(39) = -0.325
V(40) = -0.325
V(41) = -0.325
V(42) = -0.325
V(43) = -0.100
V(44) = -0.100
V(45) = -1.000
V(46) = -0.100
V(47) = -0.190
V(48) = -0.100
V(49) = -0.190
V(50) = -0.100
V(51) = -0.190
V(52) = -0.100
V(53) = -0.190
V(54) = -0.100
V(55) = -0.100
V(56) = -0.190
V(57) = -0.190
V(58) = -0.190
V(59) = -0.100
V(60) = -0.190
V(61) = -0.325
V(62) = -0.100
V(63) = -0.325
V(64) = -0.100
V(65) = -0.325
V(66) = -0.100
V(67) = -1.000
V(68

Ahora se volverá a la politica de siempre adelante y se probarán otros valores de gamma:

In [14]:
gamma = 0.95
H = 2
for i in es.policy: 
    es.policy[i]='Ad'
es.state_values=es.init_values()
es.update_values = MethodType(update_values, es)
es.solve_dynamic_programming(gamma=gamma, horizon=H)

for state in es.states:
    print(f'V({state}) = {es.state_values[state]:.3f}')

V(1) = -0.195
V(2) = -0.195
V(3) = -0.195
V(4) = -0.195
V(5) = -0.195
V(6) = -0.195
V(7) = -0.195
V(8) = -0.100
V(9) = -0.195
V(10) = -0.195
V(11) = -0.195
V(12) = -0.195
V(13) = -0.195
V(14) = -0.195
V(15) = -0.195
V(16) = -0.195
V(17) = -0.338
V(18) = -0.338
V(19) = -0.337
V(20) = -0.337
V(21) = -0.100
V(22) = -0.337
V(23) = -1.000
V(24) = -0.195
V(25) = -0.195
V(26) = -0.195
V(27) = -0.195
V(28) = -0.195
V(29) = -0.195
V(30) = -0.195
V(31) = -0.338
V(32) = -0.338
V(33) = -0.337
V(34) = -0.337
V(35) = -0.337
V(36) = -0.337
V(37) = -1.000
V(38) = -0.195
V(39) = -0.338
V(40) = -0.338
V(41) = -0.337
V(42) = -0.337
V(43) = -0.100
V(44) = -0.100
V(45) = -1.000
V(46) = -0.100
V(47) = -0.195
V(48) = -0.100
V(49) = -0.195
V(50) = -0.100
V(51) = -0.195
V(52) = -0.100
V(53) = -0.195
V(54) = -0.100
V(55) = -0.100
V(56) = -0.195
V(57) = -0.195
V(58) = -0.195
V(59) = -0.100
V(60) = -0.195
V(61) = -0.338
V(62) = -0.100
V(63) = -0.337
V(64) = -0.100
V(65) = -0.337
V(66) = -0.100
V(67) = -1.000
V(68

Ahora con la política aleatoria:

In [15]:
gamma = 0.95
H = 2
for i in es.policy: 
    rand=np.random.uniform()
    if rand<0.5:
        es.policy[i]='Ad'
    else:
        es.policy[i]='At'
es.state_values=es.init_values()
es.update_values = MethodType(update_values, es)
es.solve_dynamic_programming(gamma=gamma, horizon=H)

for state in es.states:
    print(f'V({state}) = {es.state_values[state]:.3f}')

V(1) = -0.195
V(2) = -0.195
V(3) = -0.195
V(4) = -0.195
V(5) = -0.195
V(6) = -0.195
V(7) = -0.195
V(8) = -0.100
V(9) = -0.195
V(10) = -0.195
V(11) = -0.195
V(12) = -0.195
V(13) = -0.195
V(14) = -0.195
V(15) = -0.195
V(16) = -0.195
V(17) = -0.338
V(18) = -0.338
V(19) = -0.337
V(20) = -0.195
V(21) = -0.100
V(22) = -0.195
V(23) = -1.000
V(24) = -0.195
V(25) = -0.337
V(26) = -0.195
V(27) = -0.195
V(28) = -0.195
V(29) = -0.338
V(30) = -0.195
V(31) = -0.195
V(32) = -0.195
V(33) = -0.337
V(34) = -0.337
V(35) = -0.195
V(36) = -0.337
V(37) = -1.000
V(38) = -0.195
V(39) = -0.338
V(40) = -0.338
V(41) = -0.337
V(42) = -0.337
V(43) = -0.100
V(44) = -0.100
V(45) = -1.000
V(46) = -0.100
V(47) = -0.337
V(48) = -0.100
V(49) = -0.195
V(50) = -0.100
V(51) = -0.338
V(52) = -0.100
V(53) = -0.195
V(54) = -0.100
V(55) = -0.100
V(56) = -0.195
V(57) = -0.195
V(58) = -0.195
V(59) = -0.100
V(60) = -0.195
V(61) = -0.338
V(62) = -0.100
V(63) = -0.337
V(64) = -0.100
V(65) = -0.337
V(66) = -0.100
V(67) = -1.000
V(68

### Conclusión  Actualización de valores y programación dinámica

De esta parte del documento se puede concluir que los valores más cercanos a 0 se encuentran cerca de los puntos terminales positivos (-0.021). También cuando hay casillas que se encuentran lejos de los estados terminales tienen un valor parecido (-0.195) 

## Value iteration  $V^*$

Para esta parte, lo que se hizo fue implementar el algoritmo de *Value iteration* e implementarlo en una función para poder hacer pruebas de forma sencilla.

El algoritmo tiene la siguiente estructura:
$$V^{*}(s)=\max_{\pi} \mathbb{E}\left[ \sum_{t=0}^{H} \gamma^{t} R(s_t,a_t,s_{t+1}) | \pi, s_0=s \right]$$

El objetivo de *value iteration* es tratar de encontrar la mejor política calculando el valor de cada una de las posibles acciones en el estado actual.

In [16]:
def value_iteration(es, gamma):    
    for state in es.states:       
        q = dict.fromkeys(es.get_allowed_actions(state), 0.0) 
        
        # TO DO: Actualice los valores q para Value iteration
        if state in es.snakes:
            es.policy[state]=''
            es.state_values[state] =es.state_values[es.snakes[state]]
        elif state in es.stairs:
            es.policy[state]=''
            es.state_values[state] = es.state_values[es.stairs[state]]
        else:
            for target_action in q: #Acción deseada
                for action in es.real_actions[target_action]: #Acciones resultantes para una acción deseada

                    new_state, reward, _, done = es.step(state, action, random=False) #Prueba cada acción
                    prob = es.action_probabilities[es.real_actions[target_action].index(action)] #Probabilidad de la acción resultante

                    if(done):
                        q[target_action] += prob*reward

                    else:
                        q[target_action] += prob*(reward + gamma*es.state_values[new_state])


            es.policy[state], es.state_values[state] =  es.key_max(q)

Se procede a inicializar los valores en 0 de la función de valor para hacer las pruebas respectivas

In [17]:
es.state_values = es.init_values()


Ahora se utilizará la función anteriormente creada con un horizonte de 15 y un $\gamma$ de 0.9: 

In [19]:
H=15
gamma=0.9
es.value_iteration = MethodType(value_iteration, es)
es.solve_value_iteration(gamma=gamma, horizon=H)

Cabe resaltar que lo que se hizo es no colocar ningun valor en los estados donde hay un inicio de escalera o donde se encuentra una cabeza de serpiente. El resultado de la política luego de utilizar el algoritmo de *Value iteration* fue:

In [22]:
print(es.policy)

{1: 'Ad', 2: 'Ad', 3: 'Ad', 4: 'Ad', 5: 'Ad', 6: 'Ad', 7: 'Ad', 8: '', 9: 'Ad', 10: 'Ad', 11: 'Ad', 12: 'Ad', 13: 'Ad', 14: 'Ad', 15: 'Ad', 16: 'Ad', 17: 'Ad', 18: 'Ad', 19: 'Ad', 20: 'Ad', 21: '', 22: 'At', 23: 'Ad', 24: 'At', 25: 'At', 26: 'At', 27: 'At', 28: 'At', 29: 'At', 30: 'At', 31: 'At', 32: 'At', 33: 'At', 34: 'At', 35: 'Ad', 36: 'Ad', 37: 'Ad', 38: 'Ad', 39: 'Ad', 40: 'Ad', 41: 'Ad', 42: 'Ad', 43: '', 44: '', 45: 'Ad', 46: '', 47: 'Ad', 48: '', 49: 'Ad', 50: '', 51: 'Ad', 52: '', 53: 'Ad', 54: '', 55: '', 56: 'Ad', 57: 'Ad', 58: 'Ad', 59: '', 60: 'Ad', 61: 'Ad', 62: '', 63: 'At', 64: '', 65: 'Ad', 66: '', 67: 'Ad', 68: 'Ad', 69: '', 70: 'Ad', 71: 'Ad', 72: 'Ad', 73: '', 74: 'Ad', 75: 'Ad', 76: 'Ad', 77: 'Ad', 78: 'Ad', 79: 'Ad', 80: '', 81: 'At', 82: 'At', 83: '', 84: 'At', 85: 'At', 86: 'At', 87: 'At', 88: 'At', 89: 'Ad', 90: 'At', 91: 'At', 92: '', 93: 'Ad', 94: 'Ad', 95: '', 96: 'Ad', 97: 'Ad', 98: '', 99: 'Ad', 100: 'Ad'}


Aquí ya es posible ver el comportamiento del agente.

Por ejemplo, se tiene que la casilla **23** es un estado terminal con recompensa negativa, pero en la casilla **21** se encuetra una escalera que llega a un punto muy favorable para terminar el episodio con recompensa positiva. Por ese motivo, desde la casilla **35** se ve que la mejor acción que puede tomar el agente es *Atrás*. Porque a pesar de que hay una casilla terminal negativa cerca, la escalera en la casilla **21** tiene un mejor resultado.

También se puede ver que todas las casillas antes de la casilla **21** tienen como acción óptima *Adelante*, se puede asumir que es por lo mismo explicado anteriormente.

## Prueba Value iteration con Estados terminales aleatorios

A continuación se realizará un procedimiento parecido al numeral anterior, solo que ahora se colocarán puntos terminales de forma aleatoria y se observará el comportamiento del agente en esos casos. Se debe tener en cuenta que los estados terminales deben cumplir con lo siguiente:
* Se van a tener dos casillas azules (recompensa positiva) elegidas **aleatoriamente** en las primeras dos filas del tablero
* Se van a tener siete casillas rojas (recompensa negativa) elegidas **aleatoriamente** en las primeras nueve filas del tablero.

In [38]:
import numpy as np

# Generar dos valores aleatorios diferentes entre 1 y 20, valores dentro de las dos primeras filas del tablero
casillas_azules = np.random.choice(np.arange(1, 21), size=2, replace=False)

# Generar siete valores aleatorios diferentes entre 1 y 90, valores dentro de las nueve primeras filas del tablero
casillas_rojas = np.random.choice(np.arange(1, 91), size=7, replace=False)

print("Valores de casillas azules: ", casillas_azules, "; valores de casillas rojas: ", casillas_rojas)


Valores de casillas azules:  [10  6] ; valores de casillas rojas:  [80 34 46 83 49 58 37]


Creación del nuevo tablero:

In [39]:
es=EscalerasSerpientes(casillas_azules, casillas_rojas,-0.1)

Realizar la prueba de iteración de valor con el nuevo tablero:

In [40]:
H=15
gamma=0.9
es.value_iteration = MethodType(value_iteration, es)
es.solve_value_iteration(gamma=gamma, horizon=H)

In [41]:
print(es.policy)

{1: 'Ad', 2: 'At', 3: 'Ad', 4: 'Ad', 5: 'Ad', 6: 'Ad', 7: 'At', 8: '', 9: 'Ad', 10: 'Ad', 11: 'At', 12: 'At', 13: 'At', 14: 'At', 15: 'At', 16: 'At', 17: 'At', 18: 'At', 19: 'At', 20: 'At', 21: '', 22: 'At', 23: 'At', 24: 'At', 25: 'At', 26: 'At', 27: 'At', 28: 'At', 29: 'At', 30: 'At', 31: 'At', 32: 'At', 33: 'At', 34: 'Ad', 35: 'Ad', 36: 'Ad', 37: 'Ad', 38: 'Ad', 39: 'Ad', 40: 'Ad', 41: 'Ad', 42: 'Ad', 43: '', 44: '', 45: 'At', 46: '', 47: 'At', 48: '', 49: 'Ad', 50: '', 51: 'Ad', 52: '', 53: 'At', 54: '', 55: '', 56: 'At', 57: 'At', 58: 'Ad', 59: '', 60: 'At', 61: 'At', 62: '', 63: 'Ad', 64: '', 65: 'At', 66: '', 67: 'Ad', 68: 'Ad', 69: '', 70: 'Ad', 71: 'Ad', 72: 'Ad', 73: '', 74: 'At', 75: 'At', 76: 'At', 77: 'At', 78: 'At', 79: 'At', 80: '', 81: 'Ad', 82: 'Ad', 83: '', 84: 'At', 85: 'At', 86: 'At', 87: 'At', 88: 'At', 89: 'At', 90: 'Ad', 91: 'Ad', 92: '', 93: 'Ad', 94: 'Ad', 95: '', 96: 'At', 97: 'At', 98: '', 99: 'Ad', 100: 'Ad'}


Con el ejemplo (azules casillas 10 y 6, rojas casillas 80, 34, 46, 83, 49, 58 y 37) Se puede ver que si el agente  se encuentra en casillas anteriores a las de recompensa azul, tienden a ir hacia adelante, cuando ya están en valores más altos de 20, la acción predominante es *Atrás*, ya que quieren acercarse a los puntos azules.

Una curiosidad que se puede ver, es que el agente tiende a buscar las cabezas de las serpientes cuando ya está en valores muy alejados de la recompensa. Por ejemplo, la casilla **73** es la cabeza de una serpiente que lleva hasta la casilla **1**. Cuando el agente está por debajo de esa casilla, elije la acción *Adelante* (casilla **70**). Cuando el agente está por encima de esa casilla, el agente elije la acción *Atrás* (casilla **77**).

Esto da a entender que el agente quiere llegar lo más rápido a los puntos de casilas azules, esto buscando las cabezas de las serpientes que lo lleven a las primeras dos filas del tablero.
